### PR TITLE
yosys: 0.38 -> 0.40

### PR DESCRIPTION
**Pull request by Luflosi**
## Description of changes
https://github.com/YosysHQ/yosys/releases/tag/yosys-0.39

~~This update breaks `yosys-synlig`. I'm marking this PR as draft in the hopes that it can be fixed soon @hzeller @nbdd0121.~~

https://github.com/YosysHQ/yosys/releases/tag/yosys-0.40

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by hzeller** [comment]
Let me have a look, I think upstream synlig dealt with this issue; I'll update this thread in a few hours.

**Comment by hzeller** [comment]
Ok, I confirmed that latest synlig will not yet work with Yosys 0.39.

To not hold up this Yosys release, I am fine if synlig is marked broken for now while I work out what needs to be done to get it compatible with 0.39.

**Comment by Luflosi** [comment]
I marked yosys-synlig as broken.

**Comment by Luflosi** [comment]
Updated to Yosys 0.40.

**Comment by tpwrules** [comment]
This builds and passes tests on `x86_64-linux` and `aarch64-darwin`. Found that glasgow does not build on macOS due to a test failure due to yosys synthesizing slightly less efficiently. It probably needs to be upgraded, and/or that test skipped.

**Comment by hzeller** [comment]
I've created #318592 so that yosys upgrades are not shackled by the slow update of that plugin.
Need someone to review and merge that one.

File: pkgs/development/compilers/yosys/plugins/synlig.nix [hunk 1]
```diff
@@ -97,5 +97,6 @@ stdenv.mkDerivation (finalAttrs: {
     license     = licenses.asl20;
     maintainers = with maintainers; [ hzeller ];
     platforms   = platforms.all;
+    broken      = true; # Does not compile with yosys 0.39
```
Comment (hzeller) [inline pkgs/development/compilers/yosys/plugins/synlig.nix:100]: In #318592, I narrowed this to be more explicit with the specific version at hand, but this is also good.

**Review by hzeller — APPROVED** [review]
Looks good.

**Comment by hzeller** [comment]
This review should get finished. In the meantime Yosys is already at 0.42.
I am sorry if the Synlig confusion slowed this down.

**Comment by tpwrules** [comment]
Please update to the latest Yosys and I will review.

**Comment by pbsds** [comment]
I can merge when this is ready. The new amaranth bump i have in the works requires at least yosys version 0.40

**Comment by pbsds** [comment]
i ended up playing around with this, and made this updater for abc:
```nix
  # updates to the latest tag, but the tags in this fork are unofficial
  # and the git ref is needed by yosys
  passthru.updateScript = writeShellScript "update-abc" ''
    set -euo pipefail
    repo=${finalAttrs.src.owner}/${finalAttrs.src.repo}
    rev="$(${lib.getExe gh} api repos/$repo/tags --jq '.[0].commit.sha')"
    date="$(${lib.getExe gh} api repos/$repo/commits/$rev --jq '.commit.committer.date' | cut -dT -f1)"
    ${lib.getBin common-updater-scripts}/bin/update-source-version ''${UPDATE_NIX_ATTR_PATH:-abc-verifier} 0-unstable-$date --rev=$rev "$@"
  '';
```

**Comment by pbsds** [comment]
superseeded by https://github.com/NixOS/nixpkgs/pull/322306

**Pull request closed without merging**